primitives/election-providers/src/onchain.rs

```diff
@@ -0,0 +1,168 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An implementation of [`ElectionProvider`] that does an on-chain sequential phragmen.
+
+use sp_arithmetic::InnerOf;
+use crate::{ElectionDataProvider, ElectionProvider};
+use sp_npos_elections::{
+	ElectionResult, ExtendedBalance, IdentifierT, PerThing128, Supports, VoteWeight,
+};
+use sp_std::{collections::btree_map::BTreeMap, marker::PhantomData, prelude::*};
+
+/// Errors of the on-chain election.
+#[derive(Eq, PartialEq, Debug)]
+pub enum Error {
```

Comment:
Yeah this is future proofing, using another crate's error as the current crate's error doesn't sound good to me. 

The runtime interface I am not sure.

Reply:
IMHO this is not needed, elect error is exactly npos_election_error. If elect error needs to add a variant then it is a breaking change anyway.
This enum will just require useless destructuring.
But I'm ok to keep it.

Reply:
So the scenario was that imagine: there is an error type specific to this onchain implementation, and as you said this would add a variant. Why is this a breaking change per se? you mean breaking something in the runtime interface or in the semver manner?

Note that in the runtime we might parse and interpret this error, but we won't necessarily need to pass it to the UI as `#pallet::Error` or anything.

Reply:
adding a variant to an enum is a major breaking change in semver,
or you have to use `#[non_exhaustive]` (or similar mecanism)

Otherwise any code which match the enum exhaustively will no longer compile.

Reply:
But let's keep it, it's fine.